tnc/codec2.py

```diff
@@ -0,0 +1,205 @@
+api.FREEDV_MODE_DATAC3 = 12
+api.FREEDV_MODE_DATAC0 = 14
+
+# Return code flags for freedv_get_rx_status() function
+api.FREEDV_RX_TRIAL_SYNC = 0x1       # demodulator has trial sync
+api.FREEDV_RX_SYNC       = 0x2       # demodulator has sync
+api.FREEDV_RX_BITS       = 0x4       # data bits have been returned
+api.FREEDV_RX_BIT_ERRORS = 0x8       # FEC may not have corrected all bit errors (not all parity checks OK)
+
+api.rx_sync_flags_to_text = [
+    "----",
+    "---T",
+    "--S-",
+    "--ST",
+    "-B--",
+    "-B-T",
+    "-BS-",
+    "-BST",
+    "E---",
+    "E--T",
+    "E-S-",
+    "E-ST",
+    "EB--",
+    "EB-T",
+    "EBS-",
+    "EBST"]
+
+# audio buffer ---------------------------------------------------------
+
+class audio_buffer:
+    # a buffer of int16 samples, using a fixed length numpy array self.buffer for storage
+    # self.nbuffer is the current number of samples in the buffer
+    def __init__(self, size):
+        print("create audio_buffer: ", size)
+        self.size = size
+        self.buffer = np.zeros(size, dtype=np.int16)
+        self.nbuffer = 0
+    def push(self,samples):
+        # add samples at the end of the buffer
+        assert self.nbuffer+len(samples) <= self.size
+        self.buffer[self.nbuffer:self.nbuffer+len(samples)] = samples
+        self.nbuffer += len(samples)
+    def pop(self,size):
+        # remove samples from the start of the buffer
+        self.nbuffer -= size;
+        self.buffer[:self.nbuffer] = self.buffer[size:size+self.nbuffer]
+        assert self.nbuffer >= 0
+               
+# resampler ---------------------------------------------------------
+
+api.FDMDV_OS_48         = int(6)                                       # oversampling rate
+api.FDMDV_OS_TAPS_48K   = int(48)                                      # number of OS filter taps at 48kHz
+api.FDMDV_OS_TAPS_48_8K = int(api.FDMDV_OS_TAPS_48K/api.FDMDV_OS_48)   # number of OS filter taps at 8kHz
+api.fdmdv_8_to_48_short.argtype = [c_void_p, c_void_p, c_int]
+api.fdmdv_48_to_8_short.argtype = [c_void_p, c_void_p, c_int]
+
+class resampler:
+    # resample an array of variable length, we just store the filter memories here
+    MEM8 = api.FDMDV_OS_TAPS_48_8K
+    MEM48 = api.FDMDV_OS_TAPS_48K
+
+    def __init__(self):
+        print("create 48<->8 kHz resampler")
+        self.filter_mem8 = np.zeros(self.MEM8, dtype=np.int16)
+        self.filter_mem48 = np.zeros(self.MEM48)
+        
+        
+    def resample48_to_8(self,in48):
+        assert in48.dtype == np.int16
+        # length of input vector must be an integer multiple of api.FDMDV_OS_48
+        assert(len(in48) % api.FDMDV_OS_48 == 0)
+
+        # concat filter memory and input samples
+        in48_mem = np.zeros(self.MEM48+len(in48), dtype=np.int16)
+        in48_mem[:self.MEM48] = self.filter_mem48
+        in48_mem[self.MEM48:] = in48
+
+        # In C: pin48=&in48_mem[MEM48]
+        pin48 = byref(np.ctypeslib.as_ctypes(in48_mem), 2*self.MEM48)
```

Comment:
This line cost me three hours of my life and I want them back :grimacing: !

Ideas for refactoring this class: I think embracing the ctypes library and maybe using `create_string_buffer()` for `in48_mem` here may be the way to go.  Ctypes even has `memcpy` which makes me happy.

I'm finding Python is so "big" compared to C.  So many libraries and data types, and different ways of doing things.

Reply:
I'm feeling with you! 

I think if we want to do this, we have to think about the dtype=int16 part of numpy. If we are moving away from this and handling the audio data as a byte string, we could maybe make the code a little bit simpler. One part of confusion is the usage of numpy arrays + directly access data via ctypes.

But I would suggest, that this is a last step or maybe a part for another PR as soon as we found a working design for handling the audio part.

And yes, doing things in many different way may be easy in the beginning when starting learning python, because everything is working somehow. But  some when a point is reached, where it causes more confusion and ugly code. For this reason I was thinking about moving to another programming language. But I'm not sure if this is a good idea....

Reply:
> But I would suggest, that this is a last step or maybe a part for another PR as soon as we found a working design for handling the audio part.

Yes I agree :+1: No rush on the refactoring.